packages/react-utilities/src/hooks/useControllableState.ts

```diff
@@ -0,0 +1,102 @@
+import * as React from 'react';
+import { useConst } from './useConst';
+
+type NonUndefined<T> = T extends undefined ? never : T;
+
+type UseControllableStateOptions<State> = {
+  /**
+   * User provided default state or factory initializer
+   */
+  defaultState?: State | (() => State);
+  /**
+   * User provided controllable state, undefined state means internal state will be used
+   */
+  state: State | undefined;
+  /**
+   * Used to initialize state if all user provided states are undefined
+   */
+  initialState: State;
+};
+
+function isUndefined(state: unknown): state is undefined {
+  return typeof state === 'undefined';
+}
+
+function isFactoryDispatch<State>(newState: React.SetStateAction<State>): newState is (prevState: State) => State {
+  return typeof newState === 'function';
+}
+
+/**
+ * A useState 'like' hook that allows optional user control
+ * Useful for components which allow uncontrolled and controlled behaviours for users
+ * @returns - https://reactjs.org/docs/hooks-state.html
+ */
+export const useControllableState = <State>(
+  options: UseControllableStateOptions<State>,
+): [State, React.Dispatch<React.SetStateAction<State>>] => {
+  const isControlled = useIsControlled(options.state);
+  const initialState = isUndefined(options.defaultState) ? options.initialState : options.defaultState;
+  const [internalState, setInternalState] = React.useState<State>(initialState);
+
+  const state = isControlled ? (options.state as NonUndefined<State>) : internalState;
```

Comment:
no need to use mapped type

```suggestion
  const state = isControlled ? (options.state as State) : internalState;
```

Reply:
There's need for that. otherwise the external type being provided in the signature will include `undefined` which is an erroneous value for this hook.